File: runme/mcgen/local/datan/exasrc/e7sm.c
#include <stdio.h>
#include "datsrc.h"

/* Table of constant values */

static integer c__1 = 1;

main(void)
{
    /* Local variables */
    static integer nexp, i, j, k, n, kb, id, ks;
    static double pref, pback, p, rback, r, alref, psign, rsign, rsmin, rspls,
	          rsupr, alback, confid, alsign;
    static char *frm1000 = "%s%10.5lf%s%10.8lf%s%10.8lf\n";

/* identify program to user */
    printf("%s\n", "Program E7SM simulates experiments with few events");
    printf("%s\n\n", "and reference events.");
/* ask for input */
    printf("%s\n", "Enter number NEXP of experiments (>0):");
    printf("%s", "> ");
    scanf("%ld", &nexp);
    printf("%s\n", "Enter number N of studied objects (>>0):");
    printf("%s", "> ");
    scanf("%ld", &n);
    printf("%s\n", "Enter Poisson parameter ALSIGN for signal events (>0.):");
    printf("%s", "> ");
    scanf("%lf", &alsign);
    printf("%s\n",
           "Enter Poisson parameter ALBACK for background evts (>=0.):");
    printf("%s", "> ");
    scanf("%lf", &alback);
    printf("%s\n", "Enter Poisson parameter ALREF for reference evts (>0.):");
    printf("%s", "> ");
    scanf("%lf", &alref);
    printf("%s\n", "Enter confidence limit CONFID (0.<CONFID<1.):");
    printf("%s", "> ");
    scanf("%lf", &confid);
    psign = alsign / (double) n;
    pback = alback / (double) n;
    pref = alref / (double) n;
    rsign = psign / pref;
    rback = pback / pref;
    p = psign + pback + pref;
/* write table header */
    printf("\n%5ld%s\n", nexp, " experiments will be simulated with");
    printf(frm1000, "ALSIGN = ", alsign, " i.e. PSIGN = ", psign,
           " and RSIGN = ", rsign);
    printf(frm1000, "ALBACK = ", alback, " i.e. PBACK = ", pback,
           " and RBACK = ", rback);
    printf("%s%10.5lf%s%10.8lf\n", "ALREF  = ", alref, " i.e. PREF  = ", pref);
    printf("%s%5.3lf\n", "Limits for RSIGN are obtained at a"
           " confidence level of ", confid);
    printf("\n%s\n", "    I   KS   KB    K   ID     RSMIN     RSPLS     RSUPR");
/* loop over simulated experiments */
    for (i = 1; i <= nexp; ++i) {
	ks = 0;
	kb = 0;
	id = 0;
	for (j = 1; j <= n; ++j) {
	    rnecuy_(&r, &c__1);
	    if (r < p) {
		if (r < psign) {
		    ++ks;
		} else if (r >= psign && r < psign + pback) {
		    ++kb;
		} else {
		    ++id;
		}
	    }
	}
	k = ks + kb;
/* compute confidence limits */
	smerqs_(&k, &id, &confid, &rback, &rsmin, &rspls, &rsupr);
/* output */
        printf("%5ld%5ld%5ld%5ld%5ld%10.5lf%10.5lf%10.5lf\n",
               i, ks, kb, k, id, rsmin, rspls, rsupr);
    }
    return 0;
} /* main */

File: runme/mcgen/local/datan/exasrc/e7sm.f
      PROGRAM E7SM
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
C identify program to user
      WRITE(*,*)
     +' Program E7SM simulates experiments with few events'
      WRITE(*,*)' and reference events'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number NEPX of experiments (>0)'
      WRITE(*,*)' >'
      READ(*,*)NEXP
      WRITE(*,*)' Enter number N of studied objects (>>0)'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)
     +' Enter Poisson parameter ALSIGN for signal events (>0.)'
      WRITE(*,*)' >'
      READ(*,*)ALSIGN
      WRITE(*,*)
     +' Enter Poisson parameter ALBACK for background evts (>=0.)'
      WRITE(*,*)' >'
      READ(*,*)ALBACK
      WRITE(*,*)
     +' Enter Poisson parameter ALREF for reference evts (>0.)'
      WRITE(*,*)' >'
      READ(*,*)ALREF
      WRITE(*,*)' Enter confidence limit CONFID (0.<CONFID<1.)'
      WRITE(*,*)' >'
      READ(*,*)CONFID
      PSIGN=ALSIGN/DBLE(N)
      PBACK=ALBACK/DBLE(N)
      PREF=ALREF/DBLE(N)
      RSIGN=PSIGN/PREF
      RBACK=PBACK/PREF
      P=PSIGN+PBACK+PREF
C write table header
      WRITE(*,'(/,I5,A)')NEXP,' experiments will be simulated with'
      WRITE(*,1000)' ALSIGN = ',ALSIGN,' i.e. PSIGN = ',PSIGN,
     +' and RSIGN = ',RSIGN
      WRITE(*,1000)' ALBACK = ',ALBACK,' i.e. PBACK = ',PBACK,
     +' and RBACK = ',RBACK
      WRITE(*,1000)' ALREF  = ',ALREF,' i.e. PREF  = ',PREF
1000  FORMAT(A,F10.5,A,F10.8,A,F10.8)
      WRITE(*,'(A,A,F5.3)')' Limits for RSIGN are obtained at a',
     +' confidence level of ',CONFID
      WRITE(*,'(/,A)')
     +'    I   KS   KB    K   ID     RSMIN     RSPLS     RSUPR'
C loop over simulated experiments
      DO 20 I=1,NEXP
        KS=0
        KB=0
        ID=0
        DO 10 J=1,N
          CALL RNECUY(R,1)
          IF(R.LT.P)THEN
            IF(R.LT.PSIGN)THEN
              KS=KS+1
            ELSE IF(R.GE.PSIGN .AND. R.LT.PSIGN+PBACK)THEN
              KB=KB+1
            ELSE
              ID=ID+1
            END IF
          END IF
10      CONTINUE
        K=KS+KB
C compute confidence limits
        CALL SMERQS(K,ID,CONFID,RBACK,RSMIN,RSPLS,RSUPR)
C output
        WRITE(*,'(5I5,3F10.5)')I,KS,KB,K,ID,RSMIN,RSPLS,RSUPR
20    CONTINUE
      END
